File: built-in plugins dsp/mb-reverb-convolution.cpp
/**
 * MB Convolution
 * Category : effect
 * Type     : reverb
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : IR convolution reverb
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_REVERB_CONVOLUTION_H
#define MB_REVERB_CONVOLUTION_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbReverbConvolution : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-reverb-convolution";
    static constexpr const char* PLUGIN_NAME    = "MB Convolution";
    static constexpr const char* PLUGIN_TYPE    = "reverb";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float ir = 0.5f;  // range [0, 1]
    float predelay = 20f;  // range [0, 200]
    float mix = 0.3f;  // range [0, 1]
    };

    MbReverbConvolution() = default;
    ~MbReverbConvolution() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.ir = std::clamp(params.ir, 0f, 1f);
        params.predelay = std::clamp(params.predelay, 0f, 200f);
        params.mix = std::clamp(params.mix, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Convolution
        return input;
    }
};

#endif // MB_REVERB_CONVOLUTION_H

File: built-in plugins dsp/mb-reverb-convolution.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbReverbConvolutionPlugin: PluginDefinition = { id: 'mb-reverb-convolution', slug: 'mb-reverb-convolution', name: 'MB Convolution', category: 'effect', type: 'reverb', version: '1.0.0', description: 'IR convolution reverb', author: 'Max Booster', parameters: [{ id: 'ir', name: 'Impulse', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true }, { id: 'predelay', name: 'Pre-delay', type: 'float', defaultValue: 20, minValue: 0, maxValue: 200, automatable: true }, { id: 'mix', name: 'Mix', type: 'float', defaultValue: 0.3, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { ir: 0.5, predelay: 20, mix: 0.3 } };

export default MbReverbConvolutionPlugin;
